scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+);
+
+// type: Promise<number>
+const result = Effect.runPromise(program);
+ 
+result.then(console.log);
+// Console: 1
+```
+
+notes:
+To run effects with asynchronous computations we can use runPromise, which will return a promise of the value type of the ran Effect. Just like runSync, if the ran effect fails, the returns promise will reject.
+
+---
+
+Effects are composable...
+So compose them
+
+notes:
+Running Effects should occur at the edges of your program, when you have no other choice but to run an Effect. 
+
+While you're within Effect code, Effect provides tools for you to manipulate the values inside an Effect without needing to run it.
+
+---
+
+Chaining transformations with pipe
+
+```ts
+const result = pipe(input, func1, func2, ..., funcN)
+``` 
+
+![[pipe-mermaid.png]]
+
+notes:
+before we discuss the specifics of these transformation functions, its important to stop and explain pipe. Pipe is what allows us to seamlessly carry out one transformation function after another. It takes a value and a series of functions and applies them one at a time.
+
+This is best shown with an example
+
+---
+
+```ts
+const increment = (x: number) => x + 1
+const double = (x: number) => x * 2
+const subtractTen = (x: number) => x - 10
+ 
+const result = pipe(5, increment, double, subtractTen)
+// identical to subtractTen(double(increment(5)))
+ 
+console.log(result) // Output: 2
+```
+
+notes:
+Here we have 3 functions which each take and return a number. Using pipe we apply them one at a time, feeding the output of one as the input to the next and ending up with out final result. 
+
+Pipe allows us to use functions in a clear left to right or top to down order instead of confusing deeply nested function calls.
+
+---
+
+The pipe method:
+
+```ts
+pipe(effect, func1, func2, ...)
+```
+
+```ts
+effect.pipe(func1, func2, ...)
+```
+
+These are identical
+
+notes:
+
+For convenience the Effect type also has a pipe method which is identical to passing the effect as the first argument to the pipe function, but looks a bit cleaner. I'll use these interchangeable through the examples in this video, just know they are the same and you can use whichever way works best for you.
+
+---
+
+Transforming the value of an effect with map
+```ts
+// type: Effect<never, never, string>
+const mappedEffect = pipe(
+  Effect.succeed(1),
+  Effect.map((x: number) => String(x))
+)
+ 
+console.log(Effect.runSync(mappedEffect)) // Output: "1"
+```
+
+notes:
+the map function is simple, it applies the function passed to it to the value inside an effect, and returns a new effect with the value as the output of that function.
+
+In this example the function passed to map takes in a number, in this case the 1 value from the previous effect, and converts it to a string. When we log the result after running the effect and see the output is a string 1
+
+---
+```ts
+const mappedEffect = pipe(
+  Effect.succeed({ x: 5, y: 0 }),
+  Effect.map(({x, y}: { x: number, y: number }) =>
```

Comment:
I would always avoid specifying the types in closures otherwise the viewer will ask themselves if this is mandatory and think it is verbose